sdk/core/azure-core/tests/test_request_id_policy.py

```diff
@@ -17,12 +19,45 @@ def test_request_id_policy_fix_id():
     request_id_policy.on_request(pipeline_request)
     assert request.headers["x-ms-client-request-id"] == test_request_id
 
-def test_request_id_policy_fix_on_demand_id():
+
+@pytest.mark.parametrize("auto_request_id", [True, False])
+def test_request_id_policy_fix_on_demand_id(auto_request_id):
     """Test policy with no other policy and happy path"""
     test_request_id = 'test_request_id'
-    request_id_policy = RequestIdPolicy()
+    request_id_policy = RequestIdPolicy(auto_request_id=auto_request_id)
     request = HttpRequest('GET', 'http://127.0.0.1/')
     pipeline_request = PipelineRequest(request, PipelineContext(None))
     pipeline_request.context.options['request_id'] = test_request_id
     request_id_policy.on_request(pipeline_request)
     assert request.headers["x-ms-client-request-id"] == test_request_id
```

Comment:
This goes partway, but it's possible to cover every possible case to make sure the context value always has precedence:
```
@pytest.mark.parametrize("auto_request_id,other_request_id", [(True, None), (True, "junk"), (False, None), (False, "junk")])
 def test_request_id_policy_fix_on_demand_id(auto_request_id):
     """Test policy with no other policy and happy path"""
     test_request_id = 'test_request_id'
     request_id_policy = RequestIdPolicy()
     request_id_policy = RequestIdPolicy(auto_request_id=auto_request_id)
     if other_request_id:
         request_id_policy.set_request_id(other_request_id)
     request = HttpRequest('GET', 'http://127.0.0.1/')
     pipeline_request = PipelineRequest(request, PipelineContext(None))
     pipeline_request.context.options['request_id'] = test_request_id
     request_id_policy.on_request(pipeline_request)
     assert request.headers["x-ms-client-request-id"] == test_request_id
```

Reply:
I like @bryevdv , and I would even consider doing combination automatically between the free possible inputs:
- auto request id 
- request id in init
- request id in request

```python
from itertools import product
auto_request_id_values = (True, False, None)
request_id_init_values = ("foo", None)
request_id_req_values = ("bar", None)

In [6]: list(product(auto_request_id_values, request_id_init_values, request_id_req_values))
Out[6]:
[(True, 'foo', 'bar'),
 (True, 'foo', None),
 (True, None, 'bar'),
 (True, None, None),
 (False, 'foo', 'bar'),
 (False, 'foo', None),
 (False, None, 'bar'),
 (False, None, None),
 (None, 'foo', 'bar'),
 (None, 'foo', None),
 (None, None, 'bar'),
 (None, None, None)]
...

def test_request_id_policy_fix_on_demand_id(auto_request_id, request_id_init, request_id_req):
    kwargs = {}
    if auto_request_id is not None:
        kwargs['autorest_request_id'] = auto_request_id
    if request_id_init is not None:
        kwargs['request_id'] = auto_request_id
    request_id_policy = RequestIdPolicy(**kwargs)
    # build req
    if request_id_req is not None:
       pipeline_request.context.options['request_id'] = request_id_req

    # Here call "on_request"

    # Check expected result
    if request_id_req: # This one wins everything:
        assert request.headers['x-ms-client-request-id'] == 'bar' 
    elif etc.

```